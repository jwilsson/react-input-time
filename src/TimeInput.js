import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

import getNewValue from './utils/getNewValue';

class TimeInput extends PureComponent {
    state = {
        value: this.props.initialTime,
    };

    handleChange = (event) => {
        const { onChange } = this.props;
        const { value } = this.state;

        const newValue = getNewValue(value, event.target.value);

        if (newValue) {
            this.setState({
                value: newValue,
            });

            if (onChange) {
                onChange(event, newValue);
            }
        }
    };

    render () {
        // eslint-disable-next-line no-unused-vars
        const { initialTime, input, ...props } = this.props;
        const { value } = this.state;

        if (input) {
            return React.cloneElement(input, {
                ...props,
                onChange: this.handleChange,
                value,
            });
        }

        return (
            <input
                {...props}
                onChange={this.handleChange}
                type="text"
                value={value}
            />
        );
    }
}

TimeInput.propTypes = {
    initialTime: PropTypes.string.isRequired,
    input: PropTypes.node,
    onChange: PropTypes.func,
};

export default TimeInput;
