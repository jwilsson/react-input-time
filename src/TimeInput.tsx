import React, {
    ChangeEvent,
    ComponentPropsWithoutRef,
    ReactElement,
    useState,
} from 'react';
import PropTypes from 'prop-types';

import getNewValue from './utils/getNewValue';

export interface Props extends ComponentPropsWithoutRef<'input'> {
    initialTime: string;
    input?: ReactElement;
    onChange?: (event: ChangeEvent<HTMLInputElement>) => void;
}

const TimeInput = ({ initialTime, input, onChange, ...props }: Props) => {
    const [value, setValue] = useState(initialTime);
    const handleChange = (event: ChangeEvent<HTMLInputElement>) => {
        const newValue = getNewValue(value, event.target.value);

        if (newValue) {
            setValue(newValue);

            if (onChange) {
                onChange(event);
            }
        }
    };

    if (input) {
        return React.cloneElement(input, {
            ...props,
            onChange: handleChange,
            value,
        });
    }

    return (
        <input {...props} onChange={handleChange} type="text" value={value} />
    );
};

TimeInput.propTypes = {
    initialTime: PropTypes.string.isRequired,
    input: PropTypes.node,
    onChange: PropTypes.func,
};

export default TimeInput;
