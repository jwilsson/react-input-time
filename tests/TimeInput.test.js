import React from 'react';
import { cleanup, fireEvent, render } from '@testing-library/react';

import TimeInput from '../src/TimeInput';

afterEach(cleanup);

const setup = (props) => render(<TimeInput {...props} />);

describe('TimeInput', () => {
    it('should render a text input with default props', () => {
        const { container } = setup({
            initialTime: '13:37',
        });

        expect(container).toMatchSnapshot();
    });

    it('should render a custom input with default props', () => {
        const custom = <input type="time" />;
        const { container } = setup({
            initialTime: '13:37',
            input: custom,
        });

        expect(container).toMatchSnapshot();
    });

    it('should call the "onChange" prop when using a default input', () => {
        const onChange = jest.fn();
        const { container } = setup({
            initialTime: '13:37',
            onChange,
        });

        fireEvent.change(container.firstChild, {
            target: {
                value: '14:00',
            },
        });

        expect(onChange).toHaveBeenCalledWith(expect.any(Object), '14:00');
    });

    it('should call the "onChange" prop when using a custom input', () => {
        const onChange = jest.fn();
        const { container } = setup({
            initialTime: '13:37',
            input: <input type="time" />,
            onChange,
        });

        fireEvent.change(container.firstChild, {
            target: {
                value: '14:00',
            },
        });

        expect(onChange).toHaveBeenCalledWith(expect.any(Object), '14:00');
    });

    it('should pass on other props when using a default input', () => {
        const { container } = setup({
            initialTime: '13:37',
            placeholder: 'HH:MM',
        });

        expect(container).toMatchSnapshot();
    });

    it('should pass on other props when using a custom input', () => {
        const { container } = setup({
            initialTime: '13:37',
            input: <input type="time" />,
            name: 'foo',
        });

        expect(container).toMatchSnapshot();
    });

    it('should add the separator when only given an hour', () => {
        const onChange = jest.fn();
        const { container } = setup({
            initialTime: '13:37',
            onChange,
        });

        fireEvent.change(container.firstChild, {
            target: {
                value: '14',
            },
        });

        expect(onChange).toHaveBeenCalledWith(expect.any(Object), '14:');
    });

    it('should not call the "onChange" prop when passed something invalid', () => {
        const onChange = jest.fn();
        const { container } = setup({
            initialTime: '13:37',
            onChange,
        });

        fireEvent.change(container.firstChild, {
            target: {
                value: 'invalid',
            },
        });

        expect(onChange).not.toHaveBeenCalled();
    });
});
