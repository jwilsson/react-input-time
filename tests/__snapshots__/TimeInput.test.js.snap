// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimeInput should pass on other props when using a custom input 1`] = `
<div>
  <input
    name="foo"
    type="time"
    value="13:37"
  />
</div>
`;

exports[`TimeInput should pass on other props when using a default input 1`] = `
<div>
  <input
    placeholder="HH:MM"
    type="text"
    value="13:37"
  />
</div>
`;

exports[`TimeInput should render a custom input with default props 1`] = `
<div>
  <input
    type="time"
    value="13:37"
  />
</div>
`;

exports[`TimeInput should render a text input with default props 1`] = `
<div>
  <input
    type="text"
    value="13:37"
  />
</div>
`;
